test/functional/p2p_eviction.py

```diff
@@ -0,0 +1,141 @@
+        for _ in range(4):
+            block_peer = node.add_p2p_connection(SlowP2PDataStore())
+            current_peer += 1
+            block_peer.sync_with_ping(timeout=5)
+            best_block = node.getbestblockhash()
+            tip = int(best_block, 16)
+            best_block_time = node.getblock(best_block)['time']
+            block = create_block(tip, create_coinbase(node.getblockcount() + 1), best_block_time + 1)
+            block.solve()
+            block_peer.send_blocks_and_test([block], node, success=True)
+            self.protected_peers.add(current_peer)
+
+        # The next 5 nodes are slow-pinging peers, making them eviction candidates.
+        for _ in range(5):
+            node.add_p2p_connection(SlowP2PInterface())
+            current_peer += 1
+
+
+        #The next 4 peers send us a tx, protecting them from eviction
+        for i in range(4):
+            txpeer = node.add_p2p_connection(SlowP2PInterface())
+            current_peer += 1
+            txpeer.sync_with_ping(timeout=5)
+
+            prevtx = node.getblock(node.getblockhash(i+1), 2)['tx'][0]
+            rawtx = node.createrawtransaction(
+                inputs=[{'txid': prevtx['txid'], 'vout': 0}],
+                outputs=[{node.get_deterministic_priv_key().address: 50 - 0.00125}],
+            )
+            sigtx = node.signrawtransactionwithkey(
+                hexstring=rawtx,
+                privkeys=[node.get_deterministic_priv_key().key],
+                prevtxs=[{
+                    'txid': prevtx['txid'],
+                    'vout': 0,
+                    'scriptPubKey': prevtx['vout'][0]['scriptPubKey']['hex'],
+                }],
+            )['hex']
+            txpeer.send_message(msg_tx(FromHex(CTransaction(), sigtx)))
+            self.protected_peers.add(current_peer)
+
+        # The next 8 peers have faster pings, which will usually protect them from eviction.
+        for _ in range(8):
+            fastpeer = node.add_p2p_connection(P2PInterface())
+            current_peer += 1
+            wait_until(lambda: "ping" in fastpeer.last_message, timeout=10)
+
+        # Make sure by asking the node what the actual min pings are
+        peerinfo = node.getpeerinfo()
+        pings = [[0 for i in range(2)] for j in range(len(node.p2ps))]
+
+        # After adding 21 peers, the next one hits the maxconnection limit and triggers the eviction mechanism.
+        node.add_p2p_connection(SlowP2PInterface())
+
+        # Usually the 8 fast peers are protected. In rare case of unreliable pings,
+        # one of the slower peers might have a faster min ping though.
+        for i in range(len(peerinfo)):
+            pings[i][0] = i
+            pings[i][1] = peerinfo[i]['minping'] if 'minping' in peerinfo[i] else 1000000
+
+        sorted_pings = sorted(pings, key = lambda x:x[1])
+
+        for i in range(8):
+            self.protected_peers.add(sorted_pings[i][0])
```

Comment:
At this stage, the total number of `protected_peers` is coming out to be 15, which actually should be 16 (4 blocks, 4 txs, 8 fast ping)?
This is probably happening because the previous `protected_peers.add(current_peer)` added the peers with `index + 1`, (`current_peer` is `1` for peer `0`), and `protected_peers.add(sorted_pings[i][0])` is adding them by peer index. This is causing an overlap of the set items and thus one less peer is added to the set. 

Or I might be completely wrong and this is intended behavior.

Reply:
Nice catch, I introduced this bug in the last push! Will fix later today.

Reply:
I fixed the initialization of `current_peer`  (so that  having `current_peer=0` for peer `0`).